### PR TITLE
feat: upgrade to TailwindCSS 2

### LINKED ISSUE #202: Upgrade tailwind 2.0.1
Please upgrade to version 2.0.1 which has just been released

### PR DESCRIPTION
**Pull request by Atinux**
This PR is a breaking change since TailwindCSS 2 introduced breaking changes, see https://blog.tailwindcss.com/tailwindcss-v2 and https://tailwindcss.com/docs/upgrading-to-v2#upgrade-to-node-js-12-13-or-higher

Since it does not support Node < 12.13 we have to release this module as version 4 and forcing `node >= 12.3.0`.

Nuxt 2.14 is still running PostCSS 7 since we supports Node 8, we have a PR to drop support for it but will still support Node 10, see https://github.com/nuxt/nuxt.js/pull/8346 and https://github.com/nuxt/nuxt.js/pull/8408

When Nuxt 2.15 will be out, I will update to `tailwindcss@latest` directly and release the version 4 of this module.

In the meantime I updated the documentation to explain how to upgrade to `tailwindcss@compat`: https://tailwindcss.nuxtjs.org/setup#tailwind-2

Resolves #202

### CODE REVIEW

**Comment by germi** [comment]
Hi

unfortunatelly, installing the compat version with `tailwindcss@compat` throws errors (this, a few times)

```
Module build failed (from ./node_modules/postcss-loader/src/index.js): 
TypeError: value.charCodeAt is not a function
```

strangely enough I get this error on one machine and not on another. Node 14 installed in both.

I also attempted this with no luck: https://tailwindcss.com/docs/installation#post-css-7-compatibility-build

**Comment by vinayakkulkarni** [comment]
> Hi
> 
> unfortunatelly, installing the compat version with `tailwindcss@compat` throws errors (this, a few times)
> 
> ```
> Module build failed (from ./node_modules/postcss-loader/src/index.js): 
> TypeError: value.charCodeAt is not a function
> ```
> 
> strangely enough I get this error on one machine and not on another. Node 14 installed in both.
> 
> I also attempted this with no luck: [tailwindcss.com/docs/installation#post-css-7-compatibility-build](https://tailwindcss.com/docs/installation#post-css-7-compatibility-build)
> 
> EDIT: this is related to the TailwindUI plugin apparently. when removing it, the problem is gone.

```js
const defaultTheme = require('tailwindcss/defaultTheme');

module.exports = {
  future: {
    removeDeprecatedGapUtilities: true,
    purgeLayersByDefault: true,
  },
  theme: {
    extend: {
      fontFamily: {
        sans: ['Inter var', ...defaultTheme.fontFamily.sans],
      },
    },
  },
  variants: {
    borderColor: ['responsive', 'hover', 'focus', 'active'],
    ...defaultTheme.variants,
  },
  plugins: [
    require('@tailwindcss/forms'),
    require('@tailwindcss/typography'),
    require('@tailwindcss/aspect-ratio'),
  ],
  purge: {
    // Learn more on https://tailwindcss.com/docs/controlling-file-size/#removing-unused-css
    enabled: process.env.NODE_ENV === 'production',
    content: [
      'components/**/*.vue',
      'layouts/**/*.vue',
      'pages/**/*.vue',
      'plugins/**/*.js',
      'nuxt.config.js',
    ],
  },
};
```

I believe, you have to remove `@tailwindcss/ui` from `plugins: []` - mentioned in the docs (https://tailwindui.com/changes-for-v2)

**Comment by Atinux** [comment]
Hi @vinayakkulkarni

Thank you, do you mind opening a PR to update the docs please? :)

**Comment by Heziode** [comment]
Any progress on this next major release ?

**Comment by UltraCakeBakery** [comment]
> Any progress on this next major release ?

A pull request is ready to merge https://github.com/nuxt/nuxt.js/pull/8591 which will include the changes required for this upgrade. I hope they will merge it somewhere between now and early next month.

Is this draft ready to be merged and published as soon as nuxt 2.15 gets published?

**Comment by frankievx** [comment]
Nuxt v2.15 is out! Let's merge?

**Comment by hoektoe** [comment]
When is this upgrade going to be merged or is there another way for me to ugprade to v2 on tailwindcss

**Comment by Heziode** [comment]
@hoektoe there is remaining work to be merged on other related project before it can be merged (take a look on related PR above).
 
You can already use Tailwind 2 in your project, take a look at: https://tailwindcss.nuxtjs.org/setup#tailwind-2

**Comment by ThomOrlo** [comment]
> When is this upgrade going to be merged or is there another way for me to upgrade to v2 on tailwindcss. Great work in the nuxt community.

Yes : https://tailwindcss.nuxtjs.org/setup#upgrading-tailwind

`yarn add --dev tailwindcss@npm:@tailwindcss/postcss7-compat postcss@^7 autoprefixer@^9`

**Comment by hoektoe** [comment]
> @hoektoe there is remaining work to be merged on other related project before it can be merged (take a look on related PR above).
> 
> You can already use Tailwind 2 in your project, take a look at: https://tailwindcss.nuxtjs.org/setup#tailwind-2
> 
> > Nuxt 2.14 is still using PostCSS 7, we have a [pull request](https://github.com/nuxt/nuxt.js/pull/8346) opened to upgrade the dependencies. Once merged on `v2.15`, we will release a major version of this module to include TailwindCSS 2, see [PR#203](https://github.com/nuxt-community/tailwindcss-module/pull/203).
> > In the meantime, you can upgrade Tailwind by using the [compatibility build](https://tailwindcss.com/docs/installation#post-css-7-compatibility-build):
> > ```shell
> > yarn add --dev tailwindcss@npm:@tailwindcss/postcss7-compat postcss@^7 autoprefixer@^9
> > ```
> > 
> > 
> > Please ensure you are using `Node >= 12.13.0` in order to use version 2 of TailwindCSS.
> > We have seen an issue with NPM regarding the packages resolution (see [comment](https://github.com/nuxt-community/tailwindcss-module/issues/202#issuecomment-738442349)), we highly recommend using [Yarn V1](https://classic.yarnpkg.com/en/docs/install) when working with Nuxt 2.

Awesome , sorry I misread the docs for that section.

**Comment by productfrontenddeveloper** [comment]
coming soon ? :)

**Comment by viennv1709** [comment]
Nuxt v2.15 is out!

File: package.json [hunk 1]
```diff
@@ -28,7 +28,7 @@
     "consola": "^2.15.0",
     "defu": "^3.2.2",
     "fs-extra": "^9.0.1",
-    "tailwindcss": "^1.9.6"
+    "tailwindcss": "^2.0.1-compat"
```
Comment (sduduzog) [inline package.json:31]: With the upgrade to nuxt 2.15, will we still need to depend on the compat version of tailwindcss?

Comment (izundo-viennv) [reply package.json:31]: No, we don't. Please upgrade tailwindcss to 2.0.3

> Once the rest of your tools have added support for PostCSS 8, you can move off of the compatibility build by re-installing Tailwind and its peer-dependencies using the latest tag:

**Comment by productfrontenddeveloper** [comment]
@Atinux 
Do you intend to include this here? Or will we install external. Build times have a very fast effect. 

https://windicss.netlify.app/guide/vue.html#nuxt

**Comment by vinayakkulkarni** [comment]
WindiCSS could be an add-on and not a replacement for pure tailwindcss imho

**Comment by Zenthae** [comment]
soon ?

**Comment by vinayakkulkarni** [comment]
> Looking at it next week, was recording my talk for Nuxt 3 and VueJS Amsterdam this week. Welcome to open source ☺️

Super psyched about Nuxt 3!

**Comment by pratik149** [comment]
Eagerly waiting!

**Comment by lpj145** [comment]
https://github.com/nuxt/nuxt.js/pull/8591 everything is fine, go ahead.

**Comment by luismarquesfreire** [comment]
Is it coming already? Need this urgently

**Comment by Heziode** [comment]
To all impatient :

Don't forget that we all have 24h in a day. No more.

So, just, be patient.

Note that you can already use Tailwind 2 in your project, take a look at: https://tailwindcss.nuxtjs.org/setup#tailwind-2

> Nuxt 2.14 is still using PostCSS 7, we have a [pull request](https://github.com/nuxt/nuxt.js/pull/8346) opened to upgrade the dependencies. Once merged on `v2.15`, we will release a major version of this module to include TailwindCSS 2, see [PR#203](https://github.com/nuxt-community/tailwindcss-module/pull/203).
>
> In the meantime, you can upgrade Tailwind by using the [compatibility build](https://tailwindcss.com/docs/installation#post-css-7-compatibility-build):
>
> ```bash
> yarn add --dev tailwindcss@npm:@tailwindcss/postcss7-compat postcss@^7 autoprefixer@^9
> ```
>
> Please ensure you are using `Node >= 12.13.0` in order to use version 2 of TailwindCSS.
>
> We have seen an issue with NPM regarding the packages resolution (see [comment](https://github.com/nuxt-community/tailwindcss-module/issues/202#issuecomment-738442349)), we highly recommend using [Yarn V1](https://classic.yarnpkg.com/en/docs/install) when working with Nuxt 2.

**Comment by rxsto** [comment]
In addition to the comment above there is already a beta out for the upcoming major version 4.0.0 (`@nuxtjs/tailwindcss@4.0.0-0`). This will enable you to use tailwind v2 without the compatability dependencies. More information to be found here: https://github.com/nuxt-community/tailwindcss-module/pull/270 (I've tested it, it's running perfectly fine as of now, didn't encounter any issues yet).

**Comment by UltraCakeBakery** [comment]
> To all impatient :
> 
> Don't forget that we all have 24h in a day. No more.
> 
> So, just, be patient.
> 
> Note that you can already use Tailwind 2 in your project, take a look at: https://tailwindcss.nuxtjs.org/setup#tailwind-2

I get where you come from and I agree with anyone that thinks we should never expect anything from open source project maintainers. But above the author says that the plan was to release this when nuxt 2.15 came out. It didn't, so he moved it to "next week", which was 3 weeks ago now.

People are not impatient, they are rightfully done waiting. Not all of us get away with dropping in a beta build or injecting dependencies manually. Some teams/companies have (high) standards.

**Comment by kissu** [comment]
> > To all impatient :
> > Don't forget that we all have 24h in a day. No more.
> > So, just, be patient.
> > Note that you can already use Tailwind 2 in your project, take a look at: https://tailwindcss.nuxtjs.org/setup#tailwind-2
> 
> I get where you come from and I agree with anyone that thinks we should never expect anything from open source project maintainers. But above the author says that the plan was to release this when nuxt 2.15 came out. It didn't, so he moved it to "next week", which was 3 weeks ago now.
> 
> People are not impatient, they are rightfully done waiting. Not all of us get away with dropping in a beta build or injecting dependencies manually. Some teams/companies have (high) standards.

HI, you can hit the "fork" button and submit a PR. That way, you will provide high standards to everybody here and get a lot of love coming from us. 🤗

**Comment by RomainLanz** [comment]
Another PR that bump Tailwind to version 2 has been merged.

➡️ https://github.com/nuxt-community/tailwindcss-module/pull/270

**Comment by lpj145** [comment]
> > To all impatient :
> > Don't forget that we all have 24h in a day. No more.
> > So, just, be patient.
> > Note that you can already use Tailwind 2 in your project, take a look at: https://tailwindcss.nuxtjs.org/setup#tailwind-2
> 
> I get where you come from and I agree with anyone that thinks we should never expect anything from open source project maintainers. But above the author says that the plan was to release this when nuxt 2.15 came out. It didn't, so he moved it to "next week", which was 3 weeks ago now.
> 
> People are not impatient, they are rightfully done waiting. Not all of us get away with dropping in a beta build or injecting dependencies manually. Some teams/companies have (high) standards.

Maybe i'm impatient, but not gracefully person, we need to understand open source projects always needs brave people, so i don't think what: people business/business company side is more important or great than people involved in projects, specially open source projects, "everbody need time" - is a good phrase.

Make love not war, thanks for great job. ❤

**Comment by Atinux** [comment]
Version 4 is out 🎊 

https://github.com/nuxt-community/tailwindcss-module/releases/tag/v4.0.0

**Comment by pratik149** [comment]
> Version 4 is out 🎊
> 
> https://github.com/nuxt-community/tailwindcss-module/releases/tag/v4.0.0

Thank you so much for all the efforts you guys put into improving Nuxt and its modules. ❤🤗

**Comment by hoektoe** [comment]
> Version 4 is out 🎊
> 
> https://github.com/nuxt-community/tailwindcss-module/releases/tag/v4.0.0

Awesome work guys, thank you so much for all the effort

**Pull request closed without merging**